cdap-kubernetes/src/main/java/co/cask/cdap/k8s/runtime/KubeTwillPreparer.java

```diff
@@ -0,0 +1,522 @@
+             volumeMounts:
+             - name: app-config
+               mountPath: /etc/podinfo-app
+             - name: pod-info
+               mountPath: /etc/podinfo
+           volumes:
+           - name: app-config
+             configMap:
+               name: service-runid
+           - name: pod-info
+             downwardAPI:
+               ...
+     */
+    V1PodSpec podSpec = new V1PodSpec();
+
+    V1Volume podInfoVolume = getPodInfoVolume();
+    V1Volume configVolume = getConfigVolume(configMap);
+    podSpec.setVolumes(Arrays.asList(podInfoVolume, configVolume));
+
+    V1Container container = new V1Container();
+    container.setName("cdap-k8s");
+    container.setImage(image);
+
+    V1VolumeMount podInfoMount = new V1VolumeMount();
+    podInfoMount.setName(podInfoVolume.getName());
+    podInfoMount.setMountPath(podInfo.getPodInfoDir());
+
+    String configDir = podInfo.getPodInfoDir() + "-app";
+    V1VolumeMount configMount = new V1VolumeMount();
+    configMount.setName(configVolume.getName());
+    configMount.setMountPath(configDir);
+
+    container.setVolumeMounts(Arrays.asList(podInfoMount, configMount));
+
+    List<V1EnvVar> envVars = new ArrayList<>();
+    // assumption is there is only a single runnable
+    for (Map.Entry<String, String> envEntry : environments.values().iterator().next().entrySet()) {
+      V1EnvVar envVar = new V1EnvVar();
+      envVar.setName(envEntry.getKey());
+      envVar.setValue(envEntry.getValue());
+      envVars.add(envVar);
+    }
+    container.setEnv(envVars);
+    V1ResourceRequirements resourceRequirements = new V1ResourceRequirements();
+    Map<String, Quantity> quantityMap = new HashMap<>();
+    quantityMap.put("cpu", new Quantity(String.valueOf(vcores)));
+    quantityMap.put("memory", new Quantity(String.format("%dMi", memoryMB)));
+    resourceRequirements.setRequests(quantityMap);
+    container.setResources(resourceRequirements);
+    // when other program types are supported, there will need to be a mapping from program type to main class
+    container.setArgs(Arrays.asList("co.cask.cdap.internal.app.runtime.k8s.UserServiceProgramMain", "--env=k8s",
+                                    String.format("--appSpecPath=%s/%s", configDir, APP_SPEC),
+                                    String.format("--programOptions=%s/%s", configDir, PROGRAM_OPTIONS)));
+    podSpec.setContainers(Collections.singletonList(container));
+
+    templateSpec.setSpec(podSpec);
+    spec.setTemplate(templateSpec);
+    deployment.setSpec(spec);
+
+    LOG.trace("Creating deployment {} in Kubernetes", resourceMeta.getName());
+    deployment = appsApi.createNamespacedDeployment(kubeNamespace, deployment, "true");
+    LOG.trace("Creating deployment {} in Kubernetes", resourceMeta.getName());
+    return deployment;
+  }
+
+  /*
+           volumes:
+           - name: config-volume
+             configMap:
+               name: service-runid
```

Comment:
The name is not service-runid, right? and it actually contains appSpec and programOpts inside, right?

Reply:
right, fixed